src/circuit/num.rs

```diff
@@ -160,27 +163,27 @@ impl<E: Engine> AllocatedNum<E> {
             } else {
                 if current_run.len() > 0 {
                     // This is the start of a run of zeros, but we need
-                    // to k-ary AND against `last_run` first.
+                    // to k-ary AND against `equal_so_far` first, to see if we can already conclude a<r.
```

Comment:
I don't like this change. The AND indicates "equal so far" only for the case where the constraints associated with zero bits of the constant also hold. Note that we continue to calculate later (further to the right) Π values even if the preceding prefix of a was *not* equal to the corresponding prefix of r.

Reply:
@daira somehow I'm having a hard time following, could you explain again? And perhaps suggest a better variable name?

Reply:
Perhaps looking at the correctness proof in version 2018.0-beta-29 of the protocol spec (section A.3.2.2) will clarify. Simply put, Pi<sub>m</sub> does not indicate that a<sub>i</sub> = c<sub>i</sub> for all i ≥ m; it actually indicates that a<sub>i</sub> ≥ c<sub>i</sub> for all i ≥ m.

Reply:
I'll change to `smaller_so_far`  - the smaller referring to smaller than `r`